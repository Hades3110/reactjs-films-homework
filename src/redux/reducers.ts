import { combineReducers } from "redux";
import filmListPageCountReducer from "./filmListCountAndCategories/reducer";
import filmListSortReducer from "./filmListSort/reducer";
import mainBannerReducer from "./mainBanner/reducer";

const rootReducer = combineReducers({
    pageCounter: filmListPageCountReducer,
    sortType: filmListSortReducer,
    mainBanner: mainBannerReducer
})

export default rootReducer
