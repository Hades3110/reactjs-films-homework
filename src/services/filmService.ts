interface FilmApiInterface {
    page: number
    results: FilmInterface[]
    total_pages: number
    total_results: number
}

export interface FilmInterface {
    id: number
    title: string
    poster_path: string
    vote_average: number
    overview: string
    genre_ids: []
    genres: GenreInterface[],
    key: ''
}

export interface GenreInterface {
    id: number
    name: string
}

export interface FilmInfoInterface extends FilmApiInterface {
    id: number
    backdrop_path: string
    title: string
    genres: GenreInterface[]
    vote_average: number
    runtime: number
    overview: string
}

// get API
const getResource = (url: string): Promise<FilmInfoInterface> => {
    return new Promise((resolve, reject) => {
        fetch(url)
            .then(res => resolve(res.json()))
            .catch(err => {
                reject(err)
                console.log(1)
            })
    })
}

export const getFilmlist = (categories: string, page: number) => {
    return getResource(`${process.env.API_BASE}movie/${categories}${process.env.API_KEY}&language=en-US&page=${page}`)
}

export const getSearchResult = (search: string) => {
    return getResource(`${process.env.API_BASE}search/multi${process.env.API_KEY}&language=en-US&query=${search}&page=1`)
}

export const getGenres = () => {
    return getResource(`${process.env.API_BASE}genre/movie/list${process.env.API_KEY}`)
}

export const getHeadFilm = (id: number) => {
    return getResource(`${process.env.API_BASE}movie/${id}=${process.env.API_KEY}`)
}

export const getFilmVideo = (id: number) => {
    return getResource(`${process.env.API_BASE}movie/${id}/videos${process.env.API_KEY}&language=en-US`)
}

export const getRecommendations = (id: number) => {
    return getResource(`${process.env.API_BASE}movie/${id}/recommendations${process.env.API_KEY}&language=en-US`)
}