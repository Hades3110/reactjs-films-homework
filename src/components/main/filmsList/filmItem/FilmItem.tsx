import { FilmPropertyInterface } from './types'
import { GenreInterface } from '../../../../services/filmService'
// import imageNorFound from '/public/assets/imageNotFound.png'
import iconPlay from '/public/assets/icon_play.png'
import cx from 'classnames'
import styles from './filmItem.module.scss'
import { useDispatch } from 'react-redux'
import { changeVideoWindow } from '../../../../redux/filmVideoPlay/action'
import { useState } from 'react'
import { useNavigate } from 'react-router'

const FilmItem = ({ id, title, image, rate, overview, filmGenres, sortType = true, genres }: FilmPropertyInterface) => {

    const [isColumn, setIsColumn] = useState<boolean>(false)

    const genreName: GenreInterface[] = genres.filter((el: GenreInterface) => {
        return filmGenres.find((id: number) => id === el.id)
    })
    const dispatch = useDispatch()
    const imageUrl = image ? `https://image.tmdb.org/t/p/original/${image}` : "https://i.ibb.co/wB4hZKX/image-Not-Found.png"
    const vote = rate ? +(rate / 2).toFixed(1) : 0

    const navigate = useNavigate()

    return (
        <div className={sortType ? cx(styles.filmItem, styles.row) : cx(styles.filmItem, styles.column)} onClick={() => navigate(`/movie/${id}`)}>
            <div className={styles.filmImage} style={{ backgroundImage: `url(${imageUrl})` }}>
                {isColumn ? <div className={styles.filmImage__hover__overview}>
                    <div className={styles.close} onClick={(event) => {
                        event.stopPropagation()
                        setIsColumn(false)
                    }}>&#x2715;</div>
                    <div className={styles.text}>{overview}</div>
                    <button>More</button>
                </div> :
                    <div className={styles.filmImage__hover}>
                        <img
                            src={iconPlay}
                            alt="Icon Play"
                            className={styles.filmImage__hover__icon}
                            onClick={(event) => {
                                event.stopPropagation()
                                dispatch(changeVideoWindow(true, id))
                            }} />
                        <button className={styles.filmImage__hover__button} onClick={(event) => {
                            event.stopPropagation()
                            setIsColumn(true)
                        }}>View Info</button>
                    </div>}
            </div>
            <div className={cx(styles.filmItem__title, styles.filmTitle)}>
                <span className={cx(styles.filmItem__title__name, styles.filmName)}>{title}</span>
                <span className={cx(styles.filmItem__title__rating, styles.filmRating)}>{vote}</span>
                <div className={cx(styles.filmItem__title__genres, styles.filmGenre)}>
                    {genreName.map((el: GenreInterface) => {
                        return <span className={styles.filmItem__title__genres__el} key={el.id}>{el.name}</span>
                    })}
                </div>
                {sortType ? null : <div className={styles.filmOverview}>
                    <span>Overview</span>
                    <span>{overview}</span>
                </div>
                }
            </div>
        </div>
    )
}

export default FilmItem
