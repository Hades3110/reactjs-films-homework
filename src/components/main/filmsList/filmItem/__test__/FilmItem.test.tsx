import { create } from "react-test-renderer";
import { BrowserRouter as Router } from 'react-router-dom'
import FilmItem from '../FilmItem';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import rootReducer from '../../../../../redux/reducers';
import { FilmInfoInterface } from '../../../../../services/filmService';

describe('FilmItem Component', () => {

  const initialState: FilmInfoInterface = {
    id: 0,
    backdrop_path: '',
    page: 1,
    genres: [],
    title: '',
    results: [],
    total_pages: 1,
    total_results: 1,
    vote_average: 1,
    runtime: 1,
    overview: '',
    homepage: '',
    cast: [],
  }

  const store = createStore(rootReducer)

  it('FilmItem snapshot test', () => {
    const component = create(<Provider store={store}>
      <Router>
        <FilmItem
          id={initialState.id}
          title={initialState.title}
          genres={initialState.genres}
          rate={0}
          filmGenres={[]}
          image={''}
          overview={initialState.overview}
        />
      </Router>
    </Provider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});