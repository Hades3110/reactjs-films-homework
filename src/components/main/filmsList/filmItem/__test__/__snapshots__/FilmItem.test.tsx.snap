// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmItem Component FilmItem snapshot test 1`] = `
<div
  className="filmItem row"
  onClick={[Function]}
>
  <div
    className="filmImage"
    style={
      Object {
        "backgroundImage": "url(test-file-stub)",
      }
    }
  >
    <div
      className="filmImage__hover"
    >
      <img
        alt="Icon Play"
        className="filmImage__hover__icon"
        onClick={[Function]}
        src="test-file-stub"
      />
      <button
        className="filmImage__hover__button"
        onClick={[Function]}
      >
        View Info
      </button>
    </div>
  </div>
  <div
    className="filmItem__title filmTitle"
  >
    <span
      className="filmItem__title__name filmName"
    >
      
    </span>
    <span
      className="filmItem__title__rating filmRating"
    >
      0
    </span>
    <div
      className="filmItem__title__genres filmGenre"
    />
  </div>
</div>
`;
